src/krate.rs

```diff
@@ -288,38 +290,65 @@ impl Crate {
+                                      INNER JOIN crate_owners
+                                         ON crate_owners.owner_id = teams.id
+                                      WHERE crate_owners.crate_id = $1
+                                        AND crate_owners.deleted = FALSE
+                                        AND crate_owners.owner_kind = $2"));
+        let team_rows = try!(stmt.query(&[&self.id, &(OwnerKind::Team as i32)]));
+
+        let mut owners = vec![];
+        owners.extend(user_rows.iter().map(|r| Owner::User(Model::from_row(&r))));
+        owners.extend(team_rows.iter().map(|r| Owner::Team(Model::from_row(&r))));
+        Ok(owners)
     }
 
-    pub fn owner_add(&self, conn: &Connection, who: i32,
+    pub fn owner_add(&self, app: &App, conn: &Connection, req_user: &User,
                      name: &str) -> CargoResult<()> {
-        let user = try!(User::find_by_login(conn, name).map_err(|_| {
-            human(format!("could not find user with login `{}`", name))
-        }));
+
+        let owner = match Owner::find_by_name(conn, name) {
+            Ok(owner@Owner::User(_)) => { owner }
+            Ok(Owner::Team(team)) => if try!(team.contains_user(app, req_user)) {
+                Owner::Team(team)
+            } else {
+                return Err(human(format!("only members of {} can add it as an owner", name)));
+            },
+            Err(err) => if name.contains(":") {
+                Owner::Team(try!(Team::create(app, conn, name, req_user)))
+            } else {
+                return Err(err);
+            },
+        };
+
         try!(conn.execute("INSERT INTO crate_owners
-                           (crate_id, user_id, created_at, updated_at,
-                            created_by, deleted)
-                           VALUES ($1, $2, $3, $3, $4, FALSE)",
-                          &[&self.id, &user.id, &::now(), &who]));
+                           (crate_id, owner_id, created_at, updated_at,
+                            created_by, owner_kind, deleted)
+                           VALUES ($1, $2, $3, $3, $4, $5, FALSE)",
+                          &[&self.id, &owner.id(), &::now(), &req_user.id, &owner.kind()]));
         Ok(())
     }
 
-    pub fn owner_remove(&self, conn: &Connection, _who: i32,
+    pub fn owner_remove(&self, conn: &Connection, _req_user: &User,
                         name: &str) -> CargoResult<()> {
-        let user = try!(User::find_by_login(conn, name).map_err(|_| {
+        let owner = try!(Owner::find_by_name(conn, name).map_err(|_| {
             human(format!("could not find user with login `{}`", name))
         }));
         try!(conn.execute("UPDATE crate_owners
                               SET deleted = TRUE, updated_at = $1
-                            WHERE crate_id = $2 AND user_id = $3",
-                          &[&::now(), &self.id, &user.id]));
+                            WHERE crate_id = $2 AND owner_id = $3 AND owner_kind = $4",
```

Comment:
You shouldn't need the `owner_kind` field here, `owner_id` should be unique and should directly imply what `owner_kind` should be